sample-programs/erc721/erc721.scm

```diff
@@ -0,0 +1,131 @@
+;;  the Free Software Foundation, either version 3 of the License or
+;;  (at your option) any later version.
+
+;;  Blockstack is distributed in the hope that it will be useful,
+;;  but WITHOUT ANY WARRANTY, including without the implied warranty of
+;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;;  GNU General Public License for more details.
+
+;;  You should have received a copy of the GNU General Public License
+;;  along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+
+;; Storage
+(define-map tokens-owner
+  ((token-id int)) 
+  ((owner principal)))
+(define-map tokens-spender
+  ((token-id int)) 
+  ((spender principal)))
+(define-map tokens-count
+  ((owner principal))
+  ((count int)))
+(define-map accounts-operator
+  ((operator principal) (account principal))
+  ((is-approved bool)))
+
+;; Internals
+
+;; Gets the amount of tokens owned by the specified address.
+(define (balance-of (account principal))
+  (let ((balance
+      (get count 
+        (fetch-entry tokens-count (tuple (owner account))))))
+    (if (eq? balance 'null) 0 balance)))
+
+;; Gets the owner of the specified token ID.
+(define (owner-of (token-id int)) 
+  (get owner 
+    (fetch-entry tokens-owner (tuple (token-id token-id)))))
+
+;; Gets the approved address for a token ID, or zero if no address set (approved method in ERC721)
+(define (is-spender-approved (spender principal) (token-id int))
+  (let ((spender
+      (get spender 
+        (fetch-entry tokens-spender (tuple (token-id token-id))))))
+    (if (eq? spender 'null) 'false 'true)))
+
+;; Tells whether an operator is approved by a given owner (isApprovedForAll method in ERC721)
+(define (is-operator-approved (account principal) (operator principal))
+  (let ((is-approved
+      (get is-approved 
+        (fetch-entry accounts-operator (tuple (operator operator) (account account))))))
+    (if (eq? is-approved 'null) 'false is-approved)))
+
+;; Returns whether the given actor can transfer a given token ID.
+;; To be optimized
+(define (can-transfer (actor principal) (token-id int)) 
+  (or 
+    (eq? actor (owner-of token-id)) 
+    (is-spender-approved actor token-id)
+    (is-operator-approved (owner-of token-id) actor)))
+ 
+;; Internal - Register token
+(define (register-token! (new-owner principal) (token-id int))
+  (let ((current-balance (balance-of new-owner)))
+    (begin
+      (set-entry! tokens-owner 
+        (tuple (token-id token-id))
+        (tuple (owner new-owner))) 
+      (set-entry! tokens-count 
+        (tuple (owner new-owner))
+        (tuple (count (+ 1 current-balance)))) 
+      'true)))
+
+;; Internal - Release token
+(define (release-token! (owner principal) (token-id int))
+  (let ((current-balance (balance-of owner)))
+    (begin
+      (delete-entry! tokens-spender 
+        (tuple (token-id token-id))) 
+      (set-entry! tokens-count 
+        (tuple (owner owner))
+        (tuple (count (- current-balance 1)))) 
```

Comment:
You need to check for an underflow here.  What happens if `count` becomes negative?

Reply:
We actually have 2 conditions preventing an underflow.
The first one is one line above: 
```
      (delete-entry! tokens-spender 
        (tuple (token-id token-id))) 
```
should abort the execution if the token doesn't exist.

The second one is in the `caller`: 
```
(if (can-transfer tx-sender token-id)
```